### ISSUE TITLE
High DPI versus normal DPI customisations

### ISSUE THREAD

**Issue by digitaltrails** [digitaltrails/vdu_controls#63]
While Qt does support automatically coping with High DPI and Normal DPI displays, when the application overrides some icon and pixmap sizes, the sizing then becomes fixed no matter what the DPI.  

For example, `vdu_controls` was developed on a High-DPI desktop, and some icons and the splash screen were sized specifically for that environment.  Sizes needed to be specified because the Qt defaults weren't quite what was needed.

Where `vdu_controls` explicitly sets icon and pixmaps sizes, it needs to account for High and Normal DPI in order to be well laid out in both DPI environments.

Currently, for Normal DPI:

1. The splash screen is excessively large.
2. The main screen control icons are larger than seems proportionate.
3. The Preset Dialog Activate-Edit button icons are also too large.

Where the code makes choices about sizes, it needs to test for High-DPI and make appropriate choices for Normal versus High DPI.

**Comment by digitaltrails** [comment]
When icons and pixmaps have custom sizes, the size chosen will now depending on whether a screen is Normal or High DPI.  Here are the before and after screen-captures for Normal-DPI:

![high-dpi-before](https://github.com/digitaltrails/vdu_controls/assets/5510901/a1e27fe7-c535-4753-9ed1-16768ef54acb)

![high-dpi-after](https://github.com/digitaltrails/vdu_controls/assets/5510901/561ecddb-270d-45a8-8862-1e3f75bca0f3)

**Comment by digitaltrails** [comment]
With High-DPI it's difficult to know what's appropriate, some folk may have scaled their displays, others may have just adjusted font sizes (me), and some may have done both.  I've added adjustments that make things look right for me, but that might not suit everyone, so I've also added an `adjust-for-dpi` setting that can turn off these

**Comment by digitaltrails** [comment]
Closing, will reopen if anyone raises a related problem.

**Comment by digitaltrails** [comment]
The choice of icon sizes might be better proportioned by interrogating the default font size in use.  This would have the advantage of being fully automatic with no need to be aware of DPI (the new setting could be removed).

**Comment by RokeJulianLockhart** [comment]
![Screenshot_20231019_160338](https://github.com/digitaltrails/vdu_controls/assets/42837531/a332f0be-7a17-484d-a964-d9db7a491c72) ![Screenshot_20231019_160103](https://github.com/digitaltrails/vdu_controls/assets/42837531/a4bbae16-8f4f-4d0d-89cc-b805b83a0774)

The new version now renders icons too small regardless of whether the DPI awareness option is enabled or not.

**Comment by digitaltrails** [comment]
To generalise DPI adjustments, I need to detect High DPI in a way that also to accounts for any desktop scaling.

As noted above, I thought measuring a rendered font might be a good way to find out a true measure of font versus icon size that would account for both DPI and scaling.  The approach is confirmed as workable by experimentation and by [https://forum.qt.io/topic/131666/high-dpi-scaling-detection](https://forum.qt.io/topic/131666/high-dpi-scaling-detection).

I've checked in 210be3e which bumps the version to 1.12.1 and implements this approach, in essence I just create a label and measure it's realised text height in pixels:

```
    font_height = QFontMetrics(QLabel("ABC").font()).height()
    high_dpi = font_height > 30
    log_info(f"HighDPI: {high_dpi} (standard font height={font_height})")
```
As far as I can tell most scaled displays have a general font size around 20-22 and unscaled 4K displays seem to be about 30.

This should stop high DPI icons from being used if a high-DPI desktop is scaled.  It's possible adjust-for-dpi flag is no longer necessary, but I can't be absolutely sure, so I'll leave it in place. 

I missed this issue because I don't use scaling. I don't use scaling because some photo-editing GUI's and browsers do not show the image at full DPI when scaled.

**Comment by digitaltrails** [comment]
> 0. > ![Screenshot_20231019_160103](https://github.com/digitaltrails/vdu_controls/assets/42837531/a4bbae16-8f4f-4d0d-89cc-b805b83a0774)
> 1. > ![Screenshot_20231019_160338](https://github.com/digitaltrails/vdu_controls/assets/42837531/a332f0be-7a17-484d-a964-d9db7a491c72)
> 
> The new version now renders icons too small regardless of whether the DPI awareness option is enabled or not.

Can you run the latest code in a console, it should report itself to be v1.2.1 and report what it outputs for:
```
INFO: HighDPI: True (standard font height=31)
```

I only have a 4K and a FHD monitor, so those are the only two allowed for - which two possible icon sizes.  Maybe I need to add a third size.

**Comment by digitaltrails** [comment]
The slider icon sizes are custom, either 40 or 52. 

The default bottom panel icon size appears to be 22, and I bump that to 32.

In your screen-capture, the bottom icons look to small compared to the slider too big. Do you think the slider icon is too small too?

But you also appear to be using a (custom?) fixed font, that might also have some affect.

**Comment by digitaltrails** [comment]
Maybe I should just compute icon sizes based on font height - have to derive the right algorithm though, e.g. icon_height = current_icon_height * font_hight/standard_font_height

**Comment by RokeJulianLockhart** [comment]
> https://github.com/digitaltrails/vdu_controls/issues/63#issuecomment-1771613747

<blockQuote>

```log
RokeJulianLockhart@s1e8h4:~> vdu_controls -v
22:59:21 INFO: VDU Controls 1.12.0 /usr/bin/vdu_controls  
22:59:21 INFO: python-locale: ('en_GB', 'UTF-8') Qt-locale: en_GB
22:59:21 INFO: app-style: breeze (detected a dark theme)
22:59:21 INFO: Looking for config file '/home/RokeJulianLockhart/.config/vdu_controls/vdu_controls.conf'
22:59:21 INFO: Using config file '/home/RokeJulianLockhart/.config/vdu_controls/vdu_controls.conf'
```

</blockQuote>

**Comment by digitaltrails** [comment]
Thanks, I should have given more detailed instructions.  Please download v1.12.1, I've added some logging INFO on what the program decides about DPI, which is now based on font height:

```
08:49:32 INFO: VDU Controls 1.12.1 vdu_controls.py  
08:49:32 INFO: python-locale: ('en_NZ', 'UTF-8') Qt-locale: en_GB
08:49:32 INFO: app-style: breeze (detected a light theme)
08:49:32 INFO: Looking for config file '/home/michael/.config/vdu_controls/vdu_controls.conf'
08:49:32 INFO: Using config file '/home/michael/.config/vdu_controls/vdu_controls.conf'
08:49:32 INFO: Logging to stdout
08:49:33 INFO: HighDPI: True (standard font height=31)   <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
08:49:33 INFO: Using system tray.
08:49:33 INFO: Configuring application (reconfiguring=False)...
08:49:33 INFO: ddcutil version (2, 0, 0) (dynamic-sleep=True)
08:49:38 INFO: Number of detected monitors is stable at 2 (loop=2)
08:49:38 INFO: Initializing controls for vdu_number='1' vdu_model_name='HP_ZR24w' self.vdu_stable_id='HP_ZR24w_CNT00811J6'
08:49:38 INFO: Using config file '/home/michael/.config/vdu_controls/HP_ZR24w_CNT00811J6.conf'
08:49:38 INFO: Initializing controls for vdu_number='2' vdu_model_name='LG_HDR_4K' self.vdu_stable_id='LG_HDR_4K_86395'
08:49:38 INFO: Using config file '/home/michael/.config/vdu_controls/LG_HDR_4K_86395.conf'
08:49:38 INFO: splash_message: 'DDC ID 1\nHP_ZR24w:CNT00811J6'
08:49:39 INFO: splash_message: 'DDC ID 2\nLG_HDR_4K:86395'
08:49:41 INFO: splash_message: 'Checking Presets'
08:49:41 INFO: Reading autolux file '/home/michael/.config/vdu_controls/AutoLux.conf'
08:49:41 INFO: Lux auto-brightness settings refresh - monitoring is off.
08:49:41 INFO: Scheduling is disabled
08:49:41 INFO: Completed configuring application
```
Once we know what font height (in pixels) is being used, we might be able to come up with a suitable icon size and I could use that to help code an algorithm to dynamically set the icon size.

If you want to experiment with what icon size suits your desktop, you could edit you local [vdu_controls.py](https://github.com/digitaltrails/vdu_controls/blob/210be3eb062086974861feabef04cf88711b1a19/vdu_controls.py#L2580) and change the line that sets the icon size for sliders:
```
            svg_icon.setFixedSize(52, 52) if is_high_dpi() else svg_icon.setFixedSize(40, 40)
```

**Comment by digitaltrails** [comment]
I've pushed changes to v1.12.1 that scales icons and the splash-screen based on the pixel height of the default font, the font that would be used for any Qt QLabel, which is presumably set by some desktop setting.  I think DPI is inherently expressed in the font pixel height, so the code no longer does anything based on for DPI.  

I will likely remove the new `adjust-for-dpi` setting which now does nothing.  But I need to see what this looks like on deepin, gnome, xfce, and also for a non-4k display.  I've tested on a 4K display running with high-DPI and in VirtualBox running in normal DPI - to my eyes the sizing looks OK. 

On startup, there is be a message that lists the standard font height in pixels:

```
20:01:13 INFO: standard_font_pixel_height=32
```
I actually round it up to a value divisible by two (trying to avoid fractional rounding in some of the later computation):

```
        standard_font_pixel_height = QFontMetrics(QLabel("ABC").font()).height()
        standard_font_pixel_height += standard_font_pixel_height % 2
```

There are various calls to a function called `standard_height()` in the code, should you wish to see what the scale factors applied to compute the sizes of anything, just search for it.

If a non-standard font is used as the default, I suppose it might cause issues, but I hope not.

This is tricky/messy stuff, no wonder so many applications look wrong. I also now think that allow the setting of scale-factors for the desktop doesn't provide much incentive for developers to take advantage of high-DPI.  So the poor old end user gets a poorer experience, even after investing in a nice high-DPI monitor.

**Comment by digitaltrails** [comment]
*Native_font_height* is a better name for this than *standard_font_height*.  I'm going to generalise the use of native_font_height to solve some layout issues for the different desktop DPI: 4K, FHD, 2K, etcetera.  

In many places in the code, pixel dimensions, such as minimum widget sizes, have been selected based on my desktops set font height, which is 32 pixels (4K unscaled desktop).  Many of these layouts look a bit empty on KDE FHD which defaults to 22 pixels.  Scaling dimensions in proportion to the native_font_height seems to result tighter layouts without breaking them.  I'll go down this rabbit hole and see what happens.

**Comment by digitaltrails** [comment]
BTW, in researching this topic I came across [http://invent.kde.org/plasma/plasma-desktop/-/issues/58](http://invent.kde.org/plasma/plasma-desktop/-/issues/58).  This seems to be an attempt to force a scaling only solution for aesthetic purposes. I think that's might be  a bad idea in a multi-platform (gnome, kde, kde3, ...) environment where not all applications behave properly when scaled. Aesthetic goals are only part of the picture.

**Comment by RokeJulianLockhart** [comment]
> https://github.com/digitaltrails/vdu_controls/issues/63#issuecomment-1773383497

@digitaltrails, I agree and tried to make my case there at https://invent.kde.org/plasma/plasma-desktop/-/issues/58#note_601915, but Nate was not having any of it. You could always try chipping in, but it might not do much to sway anyone making the decision.

**Comment by digitaltrails** [comment]
> > https://github.com/digitaltrails/vdu_controls/issues/63#issuecomment-1773383497
> 
> @digitaltrails, I agree and tried to make my case there at https://invent.kde.org/plasma/plasma-desktop/-/issues/58#note_601915, but Nate was not having any of it (not that it was popular, per https://invent.kde.org/plasma/plasma-desktop/-/issues/58#note_675645). You could always try chipping in, but it might not do much to sway anyone making the decision.

I think you made some good points.  I could also see Nate's got his mind made up on this.   I think it has to play out. Wait and see if it generates much push-back when it arrives on everyone's desktop.

**Comment by RokeJulianLockhart** [comment]
Yeah, probably. Considering that we're looking KDE-specific solutions instead of Qt ones, doesn't this demonstrate a rather severe lack of cohesion in the Qt framework? I feel like there should be a toolkit-wide way to call some standard icon sizes that relate to the current text size, since surely Qt itself renders both at some point in the pipeline. KDE Frameworks only has the QML option too now that http://invent.kde.org/plasma/plasma-desktop/-/issues/58 has passed.

**Comment by digitaltrails** [comment]
> Yeah, probably. Considering that we're looking KDE-specific solutions instead of Qt ones, doesn't this demonstrate a rather severe lack of cohesion in the Qt framework? I feel like there should be a toolkit-wide way to call some standard icon sizes that relate to the current text size, since surely Qt itself renders both at some point in the pipeline. KDE Frameworks only appears to have the QML option too now that http://invent.kde.org/plasma/plasma-desktop/-/issues/58 has passed, though I'm no Qt dev.

I think High DPI has kind of thrown a spanner in the works for Qt.  It doesn't quite feel right that I'm measuring text height in pixels that might not be real pixels.  There are calls I can make to find out if the display is High DPI, but then there's scaling.    It's been a while since I've done any Android coding, but at least that solidly distinguished pixels and density-independent-pixels - it made it clear what unit you're using.

**Comment by digitaltrails** [comment]
Released in [v1.20.0](https://github.com/digitaltrails/vdu_controls/tree/v1.20.0).
